src/middlewared/middlewared/plugins/nfs.py

```diff
@@ -248,36 +250,68 @@ def validate_paths(self, data, schema_name, verrors):
             verrors.add(f"{schema_name}.alldirs", "This option can only be used for datasets")
 
     @private
-    def validate_user_networks(self, other_shares, data, schema_name, verrors):
+    def validate_hosts_and_networks(self, other_shares, data, schema_name, verrors):
         dev = os.stat(data["paths"][0]).st_dev
 
-        used_networks = []
+        used_networks = defaultdict(lambda: 0)
         for share in other_shares:
             try:
                 share_dev = os.stat(share["paths"][0]).st_dev
             except Exception:
                 self.logger.warning("Failed to stat first path for %r", share, exc_info=True)
                 continue
 
-            used_networks.extend([(network, share_dev) for network in share["networks"]])
+            if share_dev == dev:
+                for host in share["hosts"]:
+                    try:
+                        host = socket.gethostbyname(host)
+                    except Exception:
+                        self.logger.warning("Unable to resolve host %r", host)
+                        continue
+
+                    try:
+                        network = ipaddress.ip_network(f"{host}/32")
+                    except Exception:
+                        self.logger.warning("Got invalid host %r", host)
+                        continue
+                    else:
+                        used_networks[network] += 1
+
+                for network in share["networks"]:
+                    try:
+                        network = ipaddress.ip_network(network, strict=False)
+                    except Exception:
+                        self.logger.warning("Got invalid network %r", network)
+                        continue
+                    else:
+                        used_networks[network] += 1
+
+                if share["alldirs"] and data["alldirs"]:
+                    verrors.add(f"{schema_name}.alldirs", "This option is only available once per mountpoint")
+
+        for i, host in enumerate(data["hosts"]):
+            try:
+                host = socket.gethostbyname(host)
```

Comment:
Am I understanding right that we are resolving hostnames twice?

Reply:
No, first we are resolving hostnames of existing shares, then we resolve hostnames of new one. Cache is still a very good idea.